### ISSUE TITLE
Trying to run on apple m1

### ISSUE THREAD

**Issue by finger92** [learning-at-home/hivemind#390]
The precompiled binary of the go-libp2p-daemon library can't be executed on apple m1. So I installed Go(1.15.15) and compiled the binary(https://github.com/learning-at-home/go-libp2p-daemon), substitute the binary file (p2pd) in hivemind path (PATH_TO_CONDA/envs/hivemind/lib/python3.9/site-packages/hivemind/hivemind_cli
). Then when I got an error when running the QuickStart program (https://learning-at-home.readthedocs.io/en/0.10.0/user/quickstart.html).

The trace is 
```
Traceback (most recent call last):
  File "/Users/yizhou/code/hivemind_demo/main.py", line 28, in <module>
    opt = hivemind.optim.DecentralizedOptimizer(
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/site-packages/hivemind/optim/simple.py", line 57, in __init__
    self.averager = TrainingAverager(
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/site-packages/hivemind/averaging/training.py", line 60, in __init__
    super().__init__(averaged_tensors=averaged_tensors, **kwargs)
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/site-packages/hivemind/averaging/averager.py", line 182, in __init__
    self.run_in_background(await_ready=True)
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/site-packages/hivemind/averaging/averager.py", line 262, in run_in_background
    self.start()
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: cannot pickle '_queue.SimpleQueue' object
Exception ignored in: <function DecentralizedOptimizer.__del__ at 0x124bad550>
Traceback (most recent call last):
  File "/Users/yizhou/miniforge3/envs/hivemind/lib/python3.9/site-packages/hivemind/optim/simple.py", line 97, in __del__
    self.stop_event.set()
AttributeError: 'DecentralizedOptimizer' object has no attribute 'stop_event'
```
The test is running on

- MacBook Pro (13-inch, M1, 2020)

- Mac OS Big Sur 11.6

- go version go1.15.15 darwin/amd64

- conda 4.10.3

- Python 3.9.6

```
pip list
Package                Version
---------------------- -----------
autopep8               1.5.7
awscli                 1.20.45
base58                 2.1.0
botocore               1.21.45
brotlipy               0.7.0
certifi                2021.5.30
cffi                   1.14.6
chardet                4.0.0
charset-normalizer     2.0.0
colorama               0.4.4
conda-package-handling 1.7.3
ConfigArgParse         1.5.2
cryptography           3.4.7
Cython                 0.29.24
docutils               0.15.2
future                 0.18.2
grpcio                 1.40.0
grpcio-tools           1.40.0
hivemind               0.10.0
idna                   3.1
jmespath               0.10.0
msgpack                1.0.2
multiaddr              0.0.9
netaddr                0.8.0
numpy                  1.21.2
olefile                0.46
Pillow                 8.3.2
pip                    21.2.4
prefetch-generator     1.0.1
protobuf               3.18.0
pyasn1                 0.4.8
pycodestyle            2.7.0
pycosat                0.6.3
pycparser              2.20
pydantic               1.8.2
pymultihash            0.8.2
pyOpenSSL              20.0.1
PySocks                1.7.1
python-dateutil        2.8.2
PyYAML                 5.4.1
requests               2.26.0
rsa                    4.7.2
ruamel-yaml-conda      0.15.80
s3transfer             0.5.0
scipy                  1.7.0
setuptools             57.4.0
six                    1.16.0
sortedcontainers       2.4.0
toml                   0.10.2
torch                  1.9.0.post2
torchvision            0.10.0a0
tqdm                   4.62.1
typing-extensions      3.10.0.2
urllib3                1.26.6
uvloop                 0.16.0
varint                 1.0.2
wheel                  0.37.0
```

Though This project suggests using docker on macOS, I hope could get some help about this problem

Btw, There are two similar error trace from other projects:
https://github.com/agronholm/apscheduler/issues/522
https://github.com/pytgcalls/pytgcalls/issues/14

**Comment by justheuristic** [comment]
Hi, @finger92 !

It would appear that the problem is mostly with python multiprocessing for macOS, so it shouldn't be specific to M1.

Currently, everyone with a macbook around me is in a mad dash running experiments to ICLR (deadline in a week and a half), so it will probably take some time before we can properly debug this.

So, in other words, we'll get it fixed, and we'll get there faster thanks to your detailed report, **but for now, please use docker :)**

**Comment by justheuristic** [comment]
There's been no comments on the issue for quite some time. If someone else wants to get hivemind working on M1, feel free to open a new issue. Kudos if you're prepared to help us resolve it :)

For instance, this is how we got hivemind working on ARM + Linux (on what seem to be Jetson AGS Xavier)
* install hivemind normally from source
* install go v 1.16.4 (<1.12 or >1.17 may not work)
* build p2pd
```bash
git clone https://github.com/learning-at-home/go-libp2p-daemon
cd go-libp2p-daemon
export GOOS=linux GOARCH=arm64 GOPATH=`pwd`/build
go build ./...
go install ./...
```
* check that it works
```bash
./build/bin/linux_arm64/p2pd --help
# ^-- should print help options, not an error message
```
* replace hivemind's default p2pd with this
```bash
PATH_TO_MY_P2PD=`python -c 'import hivemind; print(hivemind.__path__[0])'`/bin/p2pd
echo $PATH_TO_MY_P2PD
# ^-- should print a reasonable path to a binary called p2pd
mv ./build/bin/linux_arm64/p2pd $PATH_TO_MY_P2PD
```

This was possible thanks to the help of @mr_seeker from our discord community :)
